scripts/validate_rst_title_capitalization.py

```diff
@@ -0,0 +1,307 @@
+#!/usr/bin/env python
+
+"""
+GH #29641
+
+Collect the titles in the rst files and validate if they follow the proper
+capitalization convention.
+
+Prints the titles that do not follow the convention.
+
+Usage::
+./scripts/validate_rst_title_capitalization.py doc/source/development/contributing.rst
+./scripts/validate_rst_title_capitalization.py doc/source/
+
+"""
+
+import argparse
+import sys
+import re
+import os
+from os import walk
+from typing import Generator, List
+
+
+# Keynames that would not follow capitalization convention
+CAPITALIZATION_EXCEPTIONS = {
+    "pandas",
+    "Python",
+    "IPython",
+    "PyTables",
+    "Excel",
+    "JSON",
+    "HTML",
+    "SAS",
+    "SQL",
+    "BigQuery",
+    "STATA",
+    "Interval",
+    "PEP8",
+    "Period",
+    "Series",
+    "Index",
+    "DataFrame",
+    "C",
+    "Git",
+    "GitHub",
+    "NumPy",
+    "Apache",
+    "Arrow",
+    "Parquet",
+    "MultiIndex",
+    "NumFOCUS",
+    "sklearn",
+    "Docker",
+}
+
+# Lowercase representation of CAPITALIZATION_EXCEPTIONS
+CAPITALIZATION_EXCEPTIONS_LOWER = {word.lower() for word in CAPITALIZATION_EXCEPTIONS}
+
+# Dictionary of bad titles that will be printed later along with line numbers
+# Key: Document Directory, Value: Pair(Bad Title, Line Number)
+bad_title_dict = {}
+
+# Error Message:
+err_msg = "Heading capitalization formatted incorrectly. Please correctly capitalize"
+
+
+def is_following_capitalization_convention(title: str) -> bool:
+    """
+    Algorithm to determine if a heading follows the capitalization convention
+
+    This method returns true if the title follows the convention
+    and false if it does not
+
+    Parameters
+    ----------
+    title : str
+        Heading string to validate
+
+    Returns
+    -------
+    bool
+        True if capitalization is correct, False if not
+
+    """
+
+    # Remove https link if present in heading
+    title = re.sub(r"<https?:\/\/.*[\r\n]*>", "", title)
+
+    # Split with delimiters comma, semicolon and space, parentheses, colon, slashes
+    word_list = re.split(r"[;,-/():\s]\s*", title)
+
+    # Edge Case: First word is an empty string
+    if len(word_list[0]) == 0:
```

Comment:
`if not word_list[0]:`

Do you know why is this happening?

Reply:
I came across situations where the text in a heading was like the following:

'  heading'

where there was some number of spaces in the beginning of the heading.  In these situations, the split function would cause the first word in the word_list array to be an empty string.

Reply:
In this situation, should the heading be considered incorrect?